### ISSUE TITLE
updateUIView doesn't trigger loading of new events in SwiftUI and other SwiftUI-UIKit-interface issues

### ISSUE THREAD

**Issue by graldij** [kvyatkovskys/KVKCalendar#170]
I want the calendar to be updated when new elements are saved in the array `courses`: i.e. the events of the new element have to be displayed in the calendar in week-view. At the moment only closing the app after adding elements to `courses` triggers the new events to load. With the `print` statements I controlled that all data is constantly up to date (thanks to the `@Binding`) and found that a change in `courses` correctly triggers `updateUIView` to be called but actually nothing changes in the calendar. 
Is my implementation incorrect? Or should actually work an the problem is not from my side?
Any comment and suggestion is very much appreciated.

Code:

```
import SwiftUI
import KVKCalendar
import EventKit

struct CalendarRepresentable: UIViewRepresentable{
    @Binding var courses: [Courses]

    var events = [Event]()

    var calendar: CalendarView = {
        print("- calendar begins initialization")
        var style = Style()
        style.defaultType = .week
        //...
        //... other style settings
        //...
        if #available(iOS 13.0, *) {
            style.event.iconFile = UIImage(systemName: "paperclip")
        }
        print("- calendar ends initialization")
        return CalendarView(frame: UIScreen.main.bounds, style: style)

    }()


    func updateUIView(_ uiView: CalendarView, context: Context ){
        print("- updateUIView has been called. Events has \(events.count) elements. Courses has \(courses.count) elements.")

        context.coordinator.parent = self
        context.coordinator.courses = courses

        print("- calling calendar.dataSource = context.coordinator in update")
        calendar.dataSource = context.coordinator
        print("- calling calendar.delegate = context.coordinator in update")
        calendar.delegate = context.coordinator

        print("- calling reloadData in update ")
        calendar.reloadData()
        print("- calling reloadInputViews in update")
        calendar.reloadInputViews()

        print("- updateUIView is finished")

    }


    func makeUIView(context: Context) -> CalendarView {
        print("- makeUIView has been called")
        print("- calling calendar.dataSource = context.coordinator in make")
        calendar.dataSource = context.coordinator
        print("- calling calendar.delegate = context.coordinator in make")
        calendar.delegate = context.coordinator
        print("- calling reloadData in make")
        calendar.reloadData()
        print("- end of makeuiview")
        return calendar
    }


    func makeCoordinator() -> Coordinator {
        print("- makeCoordinator ")
        return Coordinator(self, courses)
    }


    class Coordinator: NSObject, CalendarDelegate, CalendarDataSource, UINavigationControllerDelegate {
        var events = [Event]()
        var parent: CalendarRepresentable
        var courses: [Courses]

        func eventsForCalendar(systemEvents: [EKEvent]) -> [Event] {
            print("- eventsForCalendar called. In Coordinator \(events.count) events and \(self.parent.courses.count) courses. Coordinator courses has \(courses.count).")
            loadEvents { (events) in
                self.events = events
            }
            print("- after eventsForCalendar: \(events.count) events")
            return self.events
        }

        init(_ parent: CalendarRepresentable, _ courses: [Courses]){
            print("- Initialize Coordinator")
            self.parent = parent
            self.courses = courses
            super.init()

            loadEvents { (events) in
                self.events = events
                self.parent.calendar.reloadData()
            }
            print("- End of Initializer Coordinator")
        }

        func eventsForCalendar() -> [Event] {
            print("- eventsForCalendar without parameter")
            return events
        }


        func loadEvents(completion: ([Event]) -> Void) {

            var events = [Event]()

            print("- loadEvents with courses having \(self.parent.courses.count) elements")

            var i: Int=0
            for course in courses {

                let isoDateStart = course.startEvent
                let dateStartFormatter = ISO8601DateFormatter()

                let isoDateEnd = course.endEvent
                let dateEndFormatter = ISO8601DateFormatter()
                var event = Event(ID: "\(i)")

                event.start = dateStartFormatter.date(from: isoDateStart)!
                event.end = dateEndFormatter.date(from: isoDateEnd)!
                event.text = course.name

                events.append(event)

                i=i+1
            }

            completion(events)
        }

    }

}
```

**Comment by kvyatkovskys** [comment]
Interesting 🤔

**Comment by kvyatkovskys** [comment]
I believe, I fixed the problem. I had a wrong example for `SwiftUI`. Try this

```
struct CalendarDisplayView: UIViewRepresentable {
    @Binding var events: [Event]
    
    private var calendar: CalendarView = {
         return CalendarView(frame: frame, style: style)
    }()
        
    func makeUIView(context: UIViewRepresentableContext<CalendarDisplayView>) -> CalendarView {
        calendar.dataSource = context.coordinator
        calendar.delegate = context.coordinator
        calendar.reloadData()
        return calendar
    }
    
    func updateUIView(_ uiView: CalendarView, context: UIViewRepresentableContext<CalendarDisplayView>) {
        context.coordinator.events = events
    }
    
    func makeCoordinator() -> CalendarDisplayView.Coordinator {
        Coordinator(self)
    }
    
   public init(events: Binding<[Event]>) {
        self._events = events
    }
    
    // MARK: Calendar DataSource and Delegate
    class Coordinator: NSObject, CalendarDataSource, CalendarDelegate {
        private let view: CalendarDisplayView
        var events: [Event] = [] {
            didSet {
                view.calendar.reloadData()
            }
        }
        
        init(_ view: CalendarDisplayView) {
            self.view = view
            super.init()
        }
        
        func eventsForCalendar(systemEvents: [EKEvent]) -> [Event] {
            return events
        }
    }
}
```

**Comment by graldij** [comment]
Thanks for your time! I have tried but now I am a bit confused about where should I now define (and possibly update) the events. Do you have any suggestion? 

Is it correct that `makeCoordinator()` doesn't return anything? Or should it be `return Coordinator(self)`?

**Comment by kvyatkovskys** [comment]
from my an example project... perhaps, it's not perfect 🙂
```
struct CalendarContentView: View {
    @State private var typeCalendar = 0
    @State var events: [Event] = []
        
    var body: some View {
        NavigationView {
            CalendarDisplayView(events: $events).edgesIgnoringSafeArea(.bottom)
                
                .navigationBarTitle("", displayMode: .inline)
                .navigationBarItems(leading:
                    HStack(alignment: .center) {
                        Picker(selection: $typeCalendar, label: Text("")) {
                            Text("Day").tag(0)
                            Text("Week").tag(1)
                            Text("Month").tag(2)
                        }
                        .pickerStyle(SegmentedPickerStyle())
                    }
                    , trailing:
                    Button(action: {
                        print("Add new event")
                        var event = Event(ID: "\(events.count + 1)")
                        if let end = Calendar.current.date(byAdding: .minute, value: 30, to: Date()) {
                            event.start = Date()
                            event.end = end
                        }
                        event.text = "event text new \(events.count + 1)"
                        events.append(event)
                    }) {
                        Image(systemName: "plus")
                    }
            )
        }
        .navigationViewStyle(StackNavigationViewStyle())
    }
}
```

**Comment by kvyatkovskys** [comment]
this function returns `Coordinator(self)`... `return` like a syntax of sugar in the last row
```
func makeCoordinator() -> CalendarDisplayView.Coordinator {
        Coordinator(self)
    }
```

**Comment by graldij** [comment]
Oh I see, you call `CalendarDisplayView` with the already updated `events` and all updating&loading is done automatically thanks to the binding. Thank you very much for the example, I will try it and let you know!
Yes you are right for the `return`, sorry for the stupid question.

**Comment by kvyatkovskys** [comment]
YW 🙂

**Comment by graldij** [comment]
Hi @kvyatkovskys , it works! Thanks again!

**Comment by graldij** [comment]
Hi @kvyatkovskys . Right now I am trying to enable the DetailView of the Events if an Event is clicked. However it is very difficult for me without having an example of that for SwiftUI. Would you be so kind to upload an example showing how the SwiftUI-UIKit is modified and how the DetailView is called in the parent View?
Thanks a lot.

**Comment by graldij** [comment]
Another issue of the integration of the UIKit calendar in SwiftUI that I get is that after reorientation of the phone (rotating it) the calendar is not automatically updated (and therefore this leads to major layout issues). Only after restarting the app the layout becoms acceptable. 
Is the error/problem from my side?
Screenshot right after rotating the phone:
![Simulator Screen Shot - iPhone 12 - 2021-03-16 at 15 23 55](https://user-images.githubusercontent.com/56401352/111325373-ee25a800-866b-11eb-88c7-3d7450ca11e6.png)
Screenshot after restarting the app
![Simulator Screen Shot - iPhone 12 - 2021-03-16 at 15 24 23](https://user-images.githubusercontent.com/56401352/111325369-ecf47b00-866b-11eb-81b6-86ba54f76314.png)

**Comment by kvyatkovskys** [comment]
Just call `calendar.reloadFrame()` when orientation is changed 🤔

**Comment by graldij** [comment]
Okay thanks. I repost here my other question if possible to answer.
> Hi @kvyatkovskys . Right now I am trying to enable the DetailView of the Events if an Event is clicked. However it is very difficult for me without having an example of that for SwiftUI. Would you be so kind to upload an example showing how the SwiftUI-UIKit is modified and how the DetailView is called in the parent View?
> Thanks a lot.

**Comment by graldij** [comment]
@kvyatkovskys did you have the chance to read my other questions?

**Comment by kvyatkovskys** [comment]
Sorry, I have a lot of work. You trying to enable `DetailView` for iPad, right?

**Comment by graldij** [comment]
Np. No, for iOS (iPhones) in weekview

**Comment by kvyatkovskys** [comment]
`detailView` works only for iPad and for `DayView`. You need to create your own detail view or use a default event controller from `EKEvent`

**Comment by graldij** [comment]
Oh ok. Thanks anyway then
